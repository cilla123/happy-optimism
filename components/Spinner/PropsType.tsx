export default interface PropsType {
  theme?: 'primary' | 'success' | 'warning' | 'error';
  size?: 'lg';
  strokeWidth?: number;
  percent?: number;
}
