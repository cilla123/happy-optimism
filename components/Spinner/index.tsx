import React, { PureComponent } from 'react';
import classnames from 'classnames';
import PropsType from './PropsType';

const DIAMETER = 62;

export interface SpinnerProps extends PropsType {
  prefixCls?: string;
  className?: string;
}

export default class Spinner extends PureComponent<SpinnerProps, any> {
  static defaultProps = {
    prefixCls: 'za-spinner',
    theme: 'primary',
    strokeWidth: 5,
    percent: 15,
  };

  render() {
    const { prefixCls, className, theme, size, percent, strokeWidth } = this.props;

    const cls = classnames(`${prefixCls}`, className, {
      [`theme-${theme}`]: !!theme,
      [`size-${size}`]: !!size,
    });

    const half = DIAMETER / 2;
    const r = half - (strokeWidth as number / 2);
    const round = 2 * Math.PI * r;
    const style = {
      strokeDasharray: `${(round * (percent as number)) / 100} ${round}`,
      strokeWidth,
    };

    return (
      <svg className={`${cls}`} viewBox={`0 0 ${DIAMETER} ${DIAMETER}`}>
        <circle className={`${prefixCls}-path`} cx={half} cy={half} r={r} fill="none" style={{ strokeWidth }} />
        <circle className={`${prefixCls}-line`} cx={half} cy={half} r={r} fill="none" style={style} />
      </svg>
    );
  }
}
