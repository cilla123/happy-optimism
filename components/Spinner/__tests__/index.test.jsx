import React from 'react';
import { render } from 'enzyme';
import toJson from 'enzyme-to-json';
import Spinner from '../index';

describe('Spinner', () => {
  it('renders correctly', () => {
    const wrapper = render(<Spinner />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
