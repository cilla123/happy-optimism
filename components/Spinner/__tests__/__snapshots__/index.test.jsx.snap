// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spinner renders correctly 1`] = `
<svg
  class="za-spinner theme-primary"
  viewBox="0 0 62 62"
>
  <circle
    class="za-spinner-path"
    cx="31"
    cy="31"
    fill="none"
    r="28.5"
    style="stroke-width:5"
  />
  <circle
    class="za-spinner-line"
    cx="31"
    cy="31"
    fill="none"
    r="28.5"
    style="stroke-dasharray:26.860617188192727 179.0707812546182;stroke-width:5"
  />
</svg>
`;
